src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -14,6 +14,48 @@ internal partial class SolutionState
     {
         private partial class CompilationTracker
         {
+            private readonly struct CompilationTrackerGeneratorInfo
+            {
+                /// <summary>
+                /// The best generated documents we have for the current state. <see cref="DocumentsAreFinal"/>
+                /// specifies whether the documents are to be considered final and can be reused, or whether they're from
+                /// a prior snapshot which needs to be recomputed.
+                /// </summary>
+                public readonly TextDocumentStates<SourceGeneratedDocumentState> Documents;
+
+                /// <summary>
+                /// The <see cref="GeneratorDriver"/> that was used for the last run, to allow for incremental reuse. May
+                /// be null if we don't have generators in the first place, haven't ran generators yet for this project,
+                /// or had to get rid of our driver for some reason.
+                /// </summary>
+                public readonly GeneratorDriver? Driver;
+
+                /// <summary>
+                /// Whether the generated documents in <see cref="Documents"/> are final and should not be regenerated. 
+                /// It's important that once we've ran generators once we don't want to run them again. Once we've ran
+                /// them the first time, those syntax trees are visible from other parts of the Workspaces model; if we
+                /// run them a second time we'd end up with new trees which would confuse our snapshot model -- once the
+                /// tree has been handed out we can't make a second tree later.
+                /// </summary>
+                public readonly bool DocumentsAreFinal;
+
+                public CompilationTrackerGeneratorInfo(
+                    TextDocumentStates<SourceGeneratedDocumentState> documents,
+                    GeneratorDriver? driver,
+                    bool documentsAreFinal)
+                {
+                    Documents = documents;
+                    Driver = driver;
+                    DocumentsAreFinal = documentsAreFinal;
+                }
+
+                public CompilationTrackerGeneratorInfo WithDocumentsAreFinal(bool documentsAreFinal)
+                    => DocumentsAreFinal == documentsAreFinal ? this : new(Documents, Driver, documentsAreFinal);
+
+                public CompilationTrackerGeneratorInfo WithDriver(GeneratorDriver? driver)
+                    => Driver == driver ? this : new(Documents, driver, DocumentsAreFinal);
```

Comment:
Make this a record so we don't have to write these?

Reply:
records don't operate this way :)  they *always* produce a new instance